src/runtime-rs/crates/runtimes/src/direct_volume/mod.rs

```diff
@@ -0,0 +1,156 @@
+use anyhow::{anyhow, Ok, Result};
+use base64;
+use serde::{Deserialize, Serialize};
+use std::fs;
+use std::path::Path;
+use std::{collections::HashMap, path::PathBuf};
+
+const KATA_DIRECT_VOLUME_ROOT_PATH: &str = "/run/kata-containers/shared/direct-volumes";
+const MOUNT_INFO_FILE_NAME: &str = "mountInfo.json";
+
+pub const DIRECT_VOLUME_STAT_URL: &str = "/direct-volume/stats";
+pub const DIRECT_VOLUME_RESIZE_URL: &str = "/direct-volume/resize";
+pub const DIRECT_VOLUME_PATH_KEY: &str = "path";
+
+#[derive(Serialize, Deserialize)]
+pub struct VolumeMountInfo {
+    pub volume_type: String,
+    pub device: String,
+    pub fs_type: String,
+    pub metadata: HashMap<String, String>,
+    pub options: Vec<String>,
+}
+#[derive(Serialize, Deserialize)]
+pub struct ResizeRequest {
+    pub size: u64,
+    pub volume_path: String,
+}
+
+fn join_path(volume_path: &String) -> PathBuf {
+    let urlencoded_vol_path = base64::encode(volume_path.as_bytes());
+    let mount_info_file_path = Path::new(KATA_DIRECT_VOLUME_ROOT_PATH).join(urlencoded_vol_path);
+    return mount_info_file_path;
+}
+
+// add writes the mount info (json string) of a direct volume into a filesystem path known to Kata Container.
+pub fn add(volume_path: &String, mount_info: &String) -> Result<()> {
+    let mount_info_dir_path = join_path(volume_path);
+    // check if path exists and is dir, if not, create this dir
+    if !mount_info_dir_path.exists() {
+        fs::create_dir_all(&mount_info_dir_path)?;
+    } else if !mount_info_dir_path.is_dir() {
+        return Err(anyhow!(
+            "{} is not a directory",
+            mount_info_dir_path.to_str().unwrap()
```

Comment:
This may be enough to display `mount_info_dir_path`' value:

```rust
anyhow!(
            "{:?} is not a directory",
            mount_info_dir_path);
```